CRM/Contact/BAO/Contact/Permission.php

```diff
@@ -307,17 +307,22 @@ public static function cacheClause($contactAlias = 'contact_a') {
     if (CRM_Core_Permission::check('view all contacts') ||
       CRM_Core_Permission::check('edit all contacts')
     ) {
-      if (is_array($contactAlias)) {
-        $wheres = [];
-        foreach ($contactAlias as $alias) {
+      if (!CRM_Core_Permission::check('access deleted contacts')) {
+        if (is_array($contactAlias)) {
```

Comment:
I'm not a big fan of code that does `if(is_array($foo)) {.. do complex stuff multiple times} else {.. do complex stuff once}` if the "complex stuff" is the same in both cases, then it's better to just cast `$foo` to an array.